aws_lambda_builders/workflows/java_gradle/DESIGN.md

```diff
@@ -47,71 +41,128 @@ Gradle DSL. This presents a similar problem to `setup.py` in the Python world in
 that arbitrary logic can be executed during build time that could affect both
 how the customer's artifact is built, and which dependencies are chosen.
 
-To solve the issue where the build environment could affect the `build.gradle`
-script, a `--use-container` option will be provided, to give the option of
-building in an environment that closely mimics the Lambda environment.
+An interesting challenge is dealing with single build and multi build projects.
+Consider the following different projects structures:
+
+**Project A**
+```
+ProjectA
+├── build.gradle
+├── gradlew
+├── src
+└── template.yaml
+```
+
+**Project B**
+```
+ProjectB
+├── common
+│   └── build.gradle
+├── lambda1
+│   └── build.gradle
+├── lambda2
+│   └── build.gradle
+├── build.gradle
+├── gradlew
+├── settings.gradle
+└── template.yaml
+```
+
+Building Project A is relatively simple since we just need to issue `gradlew
+build` and place the built ZIP within the artifact directory.
+
+Building Project B is a little more complicated. Starting at the top of the
+directory, we begin by issuing `gradlew build` as before. However, since we have
+now built multiple functions, we need *multiple* artifact directories, one for
+each function. The build workflow must have some way of mapping each function's
+artifact to the correct artifact directory.
 
 ## Implementation
 
-We leverage Gradle to do all the heavy lifting for executing the `build.gradle`
-script which will resolve and download the dependencies and build the project.
-To create the distribution ZIP, we insert a new Gradle `task` that packages the
-compiled classes, resouces, and dependencies into a ZIP.
+### Source directory, and artifact directory semantics
+
+To enable the multi build projects where a Gradle project can contain multiple
```

Comment:
This will be hidden. Using SAM CLI as an example:

A customer needing this could:
`sam build -m ./build.gradle`

SAM CLI will then conform to the details here. So we will provide the correct `source_dir` and `artifact_dir`. For the use-container, we will need to mount all of ProjectB (only way this will work, I think).

SAM CLI will only do this operation if -m is provided and the runtime is Java. Otherwise, SAM CLI will assume the Project A is what is desired (aka building the individual functions only).